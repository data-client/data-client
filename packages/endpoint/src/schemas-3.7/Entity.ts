/* eslint-disable @typescript-eslint/explicit-module-boundary-types */

import { isImmutable, denormalizeImmutable } from './ImmutableUtils.js';
import type { UnvisitFunction } from './interface.js';
import { Entity as EntitySchema } from './schema.js';
import { AbstractInstanceType } from '../normal.js';

const EmptyBase = class {} as any as abstract new (...args: any[]) => {
  pk(parent?: any, key?: string): string | undefined;
};

/**
 * Represents data that should be deduped by specifying a primary key.
 * @see https://resthooks.io/docs/api/Entity
 */
export default abstract class Entity extends EntitySchema(EmptyBase) {
  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  abstract pk(parent?: any, key?: string): string | undefined;

  /** Control how automatic schema validation is handled
   *
   * `undefined`: Defaults - throw error in worst offense
   * 'warn': only ever warn
   * 'silent': Don't bother with processing at all
   *
   * Note: this only applies to non-nested members.
   */
  protected declare static automaticValidation?: 'warn' | 'silent';

  /** Return true to merge incoming data; false keeps existing entity
   *
   * @see https://resthooks.io/docs/api/schema.Entity#useIncoming
   */
  static useIncoming(
    existingMeta: { date: number; fetchedAt: number },
    incomingMeta: { date: number; fetchedAt: number },
    existing: any,
    incoming: any,
  ) {
    return existingMeta.fetchedAt <= incomingMeta.fetchedAt;
  }

  /** Run when an existing entity is found in the store */
  static mergeWithStore(
    existingMeta:
      | {
          date: number;
          fetchedAt: number;
        }
      | undefined,
    incomingMeta: { date: number; fetchedAt: number },
    existing: any,
    incoming: any,
  ) {
    const useIncoming =
      // we may have in store but not in meta; so this existance check is still important
      !existingMeta ||
      this.useIncoming(existingMeta, incomingMeta, existing, incoming);

    if (useIncoming) {
      // distinct types are not mergeable, so just replace
      if (typeof incoming !== typeof existing) {
        return incoming;
      } else {
        return this.merge(existing, incoming);
      }
    } else {
      return existing;
    }
  }

  /** Factory method to convert from Plain JS Objects.
   *
   * @param [props] Plain Object of properties to assign.
   */
  declare static fromJS: <T extends typeof Entity>(
    this: T,
    // TODO: this should only accept members that are not functions
    props?: Partial<AbstractInstanceType<T>>,
  ) => AbstractInstanceType<T>;

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  declare static pk: <T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
  ) => string | undefined;

  /** Do any transformations when first receiving input */
  static process(input: any, parent: any, key: string | undefined): any {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      if (Array.isArray(input)) {
        const errorMessage = `Attempted to initialize ${
          this.name
        } with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: ${JSON.stringify(input.slice(0, 3), null, 2)}`;
        if (this.automaticValidation !== 'warn') {
          const error = new Error(errorMessage);
          (error as any).status = 400;
          throw error;
        }
        console.warn(errorMessage);
      }
    }

    return super.process(input, parent, key);
  }

  static validate(processedEntity: any): string | undefined {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      const keysOfRecord = new Set(Object.keys(this.defaults));
      const keysOfProps = Object.keys(processedEntity);
      const [found, missing, unexpected] = [[], [], []] as [
        string[],
        string[],
        string[],
      ];
      for (const keyOfProps of keysOfProps) {
        if (keysOfRecord.has(keyOfProps)) {
          found.push(keyOfProps);
        } else {
          unexpected.push(keyOfProps);
        }
      }
      for (const keyOfRecord of keysOfRecord) {
        if (!found.includes(keyOfRecord)) {
          missing.push(keyOfRecord);
        }
      }

      // only bother with this if they used *any* defaults
      if (keysOfRecord.size) {
        const tooManyUnexpected =
          // unexpected compared to members in response
          Math.max(keysOfProps.length / 2, 1) <= unexpected.length &&
          // unexpected compared to what we specified
          keysOfRecord.size > Math.max(unexpected.length, 2) &&
          // as we find more and more be more easily assured it is correct
          found.length ** 1.5 / 2 <= unexpected.length;
        const foundNothing = found.length < Math.min(1, keysOfRecord.size / 2);
        // if we find nothing (we expect at least one member for a pk)
        // or we find too many unexpected members
        if (tooManyUnexpected || foundNothing) {
          let extra = '';
          let reason = 'substantially different than expected keys';
          if (foundNothing) {
            extra += `\n    Missing: ${missing}`;
            reason = 'no matching keys found';
          }
          if (tooManyUnexpected) {
            extra += `\n    Unexpected keys: ${unexpected}`;
            reason = 'a large number of unexpected keys found';
          }
          const errorMessage = `Attempted to initialize ${
            this.name
          } with ${reason}

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema
  If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

  Expected keys:
    Found: ${found}${extra}
  Value (processed): ${JSON.stringify(processedEntity, null, 2)}`;
          if (
            (found.length >= 4 && tooManyUnexpected) ||
            this.automaticValidation === 'warn'
          ) {
            console.warn(errorMessage);
          } else {
            return errorMessage;
          }
        }
      }
    }
    return super.validate(processedEntity);
  }

  static denormalize<T extends typeof Entity>(
    this: T,
    input: any,
    unvisit: UnvisitFunction,
  ): [denormalized: AbstractInstanceType<T>, found: boolean, suspend: boolean] {
    // TODO: remove codecov ignore once denormalize is modified to expect this
    /* istanbul ignore if */
    if (typeof input === 'symbol') {
      return [undefined, true, true] as any;
    }
    // TODO(breaking): Remove fromJS and setLocal call once old versions are no longer supported
    if (isImmutable(input)) {
      if (this.validate((input as any).toJS()))
        return [undefined as any, false, true];
      // Need to set this first so that if it is referenced further within the
      // denormalization the reference will already exist.
      unvisit.setLocal?.(input);
      const [denormEntity, found, deleted] = denormalizeImmutable(
        this.schema,
        input,
        unvisit,
      );
      return [this.fromJS(denormEntity.toObject()) as any, true, deleted];
    }
    let entityCopy: any;
    // new path
    if (input instanceof this) {
      entityCopy = input;
      // TODO(breaking): Remove fromJS and setLocal call once old versions are no longer supported
    } else {
      if (this.validate(input)) {
        return [undefined as any, false, true];
      }
      entityCopy = this.fromJS(input);
      // Need to set this first so that if it is referenced further within the
      // denormalization the reference will already exist.
      unvisit.setLocal?.(entityCopy);
    }

    let deleted = false;

    // note: iteration order must be stable
    Object.keys(this.schema).forEach(key => {
      const schema = this.schema[key];
      const nextInput = (input as any)[key];
      const [value, , deletedItem] = unvisit(nextInput, schema);

      if (deletedItem && !!this.defaults[key]) {
        deleted = true;
      }
      if ((input as any)[key] !== value) {
        // we're cheating because we know it is implemented
        (this as any).set(entityCopy, key, value);
      }
    });

    return [entityCopy, true, deleted];
  }

  /** Used by denormalize to set nested members */
  protected static set?(entity: any, key: string, value: any) {
    entity[key] = value;
  }
}

if (process.env.NODE_ENV !== 'production') {
  /* istanbul ignore else */
  const superFrom = Entity.fromJS;
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof Entity>(
    this: T,
    props?: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === Entity.prototype.pk)
      throw new Error('cannot construct on abstract types');
    return superFrom.call(this, props) as any;
  };
}
