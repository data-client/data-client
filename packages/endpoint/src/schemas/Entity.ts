/* eslint-disable @typescript-eslint/explicit-module-boundary-types */

import { AbstractInstanceType } from '../normal.js';
import { Entity as EntitySchema } from '../schema.js';

const EmptyBase = class {} as any as abstract new (...args: any[]) => {
  pk(parent?: any, key?: string, args?: readonly any[]): string | undefined;
};

/**
 * Represents data that should be deduped by specifying a primary key.
 * @see https://dataclient.io/docs/api/Entity
 */
export default abstract class Entity extends EntitySchema(EmptyBase) {
  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   * @param [args] ...args sent to Endpoint
   * @see https://dataclient.io/rest/api/Entity#pk
   */
  abstract pk(
    parent?: any,
    key?: string,
    args?: readonly any[],
  ): string | undefined;

  /** Control how automatic schema validation is handled
   *
   * `undefined`: Defaults - throw error in worst offense
   * 'warn': only ever warn
   * 'silent': Don't bother with processing at all
   *
   * Note: this only applies to non-nested members.
   */
  protected declare static automaticValidation?: 'warn' | 'silent';

  /** Factory method to convert from Plain JS Objects.
   *
   * @param [props] Plain Object of properties to assign.
   * @see https://dataclient.io/rest/api/Entity#fromJS
   */
  declare static fromJS: <T extends typeof Entity>(
    this: T,
    // TODO: this should only accept members that are not functions
    props?: Partial<AbstractInstanceType<T>>,
  ) => AbstractInstanceType<T>;

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   * @param [args] ...args sent to Endpoint
   */
  declare static pk: <T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
    args?: any[],
  ) => string | undefined;

  /** Do any transformations when first receiving input
   *
   * @see https://dataclient.io/docs/api/Entity#process
   */
  static process(
    input: any,
    parent: any,
    key: string | undefined,
    args: any[],
  ): any {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      if (Array.isArray(input)) {
        const errorMessage = `Attempted to initialize ${
          this.name
        } with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://dataclient.io/docs/guides/debugging
Learn more about schemas: https://dataclient.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: ${JSON.stringify(input.slice(0, 3), null, 2)}`;
        if (this.automaticValidation !== 'warn') {
          const error = new Error(errorMessage);
          (error as any).status = 400;
          throw error;
        }
        console.warn(errorMessage);
      }
    }

    return super.process(input, parent, key, args);
  }

  declare static denormalize: <T extends typeof Entity>(
    this: T,
    input: any,
    args: readonly any[],
    unvisit: (input: any, schema: any) => any,
  ) => AbstractInstanceType<T>;
}

if (process.env.NODE_ENV !== 'production') {
  /* istanbul ignore else */
  const superFrom = Entity.fromJS;
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof Entity>(
    this: T,
    props?: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === Entity.prototype.pk)
      throw new Error('cannot construct on abstract types');
    return superFrom.call(this, props) as any;
  };
}
