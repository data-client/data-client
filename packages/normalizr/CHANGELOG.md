# Change Log

## 0.10.0

### Minor Changes

- [#2912](https://github.com/reactive/data-client/pull/2912) [`922be79`](https://github.com/reactive/data-client/commit/922be79169a3eeea8e336eee519c165431ead474) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING CHANGE: `null` inputs are no longer filtered from Array or Object

  - `[]` and [schema.Array](https://dataclient.io/rest/api/Array) now behave in the same manner.
  - `null` values are now consistently handled everywhere (being retained).
    - These were already being retained in [nested Entities](https://dataclient.io/rest/guides/relational-data#nesting)
  - `undefined` is still filtered out.

### Patch Changes

- [`67f4e0b`](https://github.com/reactive/data-client/commit/67f4e0b45068da32d20e250267cb1cd2cea51226) Thanks [@ntucker](https://github.com/ntucker)! - Update README

- [`053e823`](https://github.com/reactive/data-client/commit/053e82377bd29f200cd7dfbc700da7a3ad7fa8d7) Thanks [@ntucker](https://github.com/ntucker)! - Update NextJS Demo link

## 0.9.5

### Patch Changes

- [`bb4b9583c5`](https://github.com/reactive/data-client/commit/bb4b9583c52e2b2fe45765af10b385b571901ee7) Thanks [@ntucker](https://github.com/ntucker)! - docs: Update readme

## 0.9.4

### Patch Changes

- [`d1b51af7ac`](https://github.com/reactive/data-client/commit/d1b51af7ac4a8a7c0559f478cc9503be8e61514c) Thanks [@ntucker](https://github.com/ntucker)! - Fix unpkg bundles by ensuring dependencies are built in order

## 0.9.3

### Patch Changes

- [#2818](https://github.com/reactive/data-client/pull/2818) [`fc0092883f`](https://github.com/reactive/data-client/commit/fc0092883f5af42a5d270250482b7f0ba9845e95) Thanks [@ntucker](https://github.com/ntucker)! - Fix unpkg bundles and update names

  - Client packages namespace into RDC
    - @data-client/react - RDC
    - @data-client/core - RDC.Core
    - @data-client/redux - RDC.Redux
  - Definition packages namespace top level
    - @data-client/rest - Rest
    - @data-client/graphql - GraphQL
    - @data-client/img - Img
    - @data-client/endpoint - Endpoint
  - Utility
    - @data-client/normalizr - normalizr
    - @data-client/use-enhanced-reducer - EnhancedReducer

## 0.9.2

### Patch Changes

- [`4ea0bc83f6`](https://github.com/reactive/data-client/commit/4ea0bc83f65f49cb2155f6aecdc5f8d1b168fd5e) Thanks [@ntucker](https://github.com/ntucker)! - Docs: Update repo links to reactive organization

## 0.8.0

### Minor Changes

- [#2784](https://github.com/reactive/data-client/pull/2784) [`c535f6c0ac`](https://github.com/reactive/data-client/commit/c535f6c0ac915b5242c1c7694308b7ee7aab16a1) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING CHANGES:

  - DELETE removed -> INVALIDATE
  - drop all support for legacy schemas
    - entity.expiresAt removed
    - Collections.infer does entity check
    - all Entity overrides for backcompat are removed - operates just like EntitySchema, except with extra validation

- [#2795](https://github.com/reactive/data-client/pull/2795) [`79e286109b`](https://github.com/reactive/data-client/commit/79e286109b5566f8e7acfdf0f44201263072d1d1) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING: [Schema Serializers](https://dataclient.io/rest/guides/network-transform#deserializing-fields) _must_ support function calls

  This means Date will no longer work like before. Possible migrations:

  ```ts
  class Ticker extends Entity {
    trade_id = 0;
    price = 0;
    time = Temporal.Instant.fromEpochSeconds(0);

    pk(): string {
      return `${this.trade_id}`;
    }
    static key = "Ticker";

    static schema = {
      price: Number,
      time: Temporal.Instant.from,
    };
  }
  ```

  or to continue using Date:

  ```ts
  class Ticker extends Entity {
    trade_id = 0;
    price = 0;
    time = Temporal.Instant.fromEpochSeconds(0);

    pk(): string {
      return `${this.trade_id}`;
    }
    static key = "Ticker";

    static schema = {
      price: Number,
      time: (iso: string) => new Date(iso),
    };
  }
  ```

- [#2792](https://github.com/reactive/data-client/pull/2792) [`35ccedceb5`](https://github.com/reactive/data-client/commit/35ccedceb53d91dd54dd996990c7c75719be2b85) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING: Serializer schemas are only processed during denormalization

### Patch Changes

- [#2779](https://github.com/reactive/data-client/pull/2779) [`ff51e71f45`](https://github.com/reactive/data-client/commit/ff51e71f45857eb172f3fe05829e34c9abb68252) Thanks [@ntucker](https://github.com/ntucker)! - Update jsdocs references to dataclient.io

## 0.2.2

### Patch Changes

- 7b835f113a: Improve package tags

## 0.2.1

### Patch Changes

- 954e06e839: docs: Update README

## 0.2.0

### Minor Changes

- bf141cb5a5: Removed deprecated Endpoint.optimisticUpdate -> use Endpoint.getOptimisticResponse

### Patch Changes

- 87475a0cae: Mark expiresAt and useIncoming as deprecated in the public interface (still fine internally)
