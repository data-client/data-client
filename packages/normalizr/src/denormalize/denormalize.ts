import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import type { Denormalize, DenormalizeNullable } from '../types.js';
import { getEntities } from '../WeakEntityMap.js';

export const denormalize = <S extends Schema>(
  input: any,
  schema: S | undefined,
  entities: any,
): Denormalize<S> | DenormalizeNullable<S> => {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(getEntities(entities), new LocalCache())(input, schema)[0];
};
