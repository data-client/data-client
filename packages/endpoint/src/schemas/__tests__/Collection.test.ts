// eslint-env jest
import { inferResults, normalize, WeakEntityMap } from '@rest-hooks/normalizr';
import { IDEntity } from '__tests__/new';
import { fromJS, Record } from 'immutable';

import { denormalizeSimple as denormalize } from './denormalize';
import WeakListMap from './legacy-compat/WeakListMap';
import { AbstractInstanceType } from '../..';
import { schema } from '../..';
import Entity from '../Entity';

let dateSpy: jest.SpyInstance;
beforeAll(() => {
  dateSpy = jest
    // eslint-disable-next-line no-undef
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

const values = <T extends { [k: string]: any }>(obj: T) =>
  Object.keys(obj).map(key => obj[key]);

class Tacos extends IDEntity {
  readonly name: string = '';
  readonly alias: string | undefined = undefined;
}

class Todo extends IDEntity {
  userId = 0;
  title = '';
  completed = false;

  static key = 'Todo';
}

class User extends IDEntity {
  name = '';
  username = '';
  email = '';
  todos: Todo[] = [];

  static key = 'User';
  static schema = {
    todos: new schema.Collection(new schema.Array(Todo), {
      nestKey: (parent, key) => ({
        userId: parent.id,
      }),
    }),
  };
}

const userTodos = new schema.Collection(new schema.Array(Todo), {
  argsKey: ({ userId }: { userId: string }) => ({
    userId,
  }),
});

describe(`${schema.Collection.name} normalization`, () => {
  let warnSpy: jest.SpyInstance;
  afterEach(() => {
    warnSpy.mockRestore();
  });
  beforeEach(() => (warnSpy = jest.spyOn(console, 'warn')));

  test('throws without a key option', () => {
    // @ts-expect-error
    expect(() => new schema.Collection(new schema.Array(Todo), {})).toThrow();
  });

  test('should throw a custom error if data loads with string unexpected value', () => {
    function normalizeBad() {
      normalize('abc', userTodos);
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('should throw a custom error if data loads with string unexpected value', () => {
    function normalizeBad() {
      normalize(null, userTodos.push);
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('should throw a custom error if data loads with string unexpected value', () => {
    function normalizeBad() {
      // @ts-expect-error
      userTodos.normalize(
        [],
        undefined,
        '',
        () => undefined,
        () => undefined,
        {},
      );
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('normalizes nested collections', () => {
    const state = normalize(
      {
        id: '1',
        username: 'bob',
        todos: [{ id: '5', title: 'finish collections' }],
      },
      User,
    );
    expect(state).toMatchSnapshot();
    //const a: string | undefined = state.result;
    // @ts-expect-error
    const b: Record<any, any> | undefined = state.result;
  });

  test('normalizes top level collections', () => {
    const state = normalize(
      [{ id: '5', title: 'finish collections' }],
      userTodos,
      [{ userId: '1' }],
    );
    expect(state).toMatchSnapshot();
    //const a: string[] | undefined = state.result;
    // @ts-expect-error
    const b: Record<any, any> | undefined = state.result;
  });

  test('normalizes already processed entities', () => {
    const state = normalize(
      {
        id: '1',
        username: 'bob',
        todos: ['5', '6'],
      },
      User,
    );
    expect(state).toMatchSnapshot();
  });

  test('normalizes push onto the end', () => {
    const init = {
      entities: {
        'COLLECT:ArraySchema(Todo)': {
          '{"userId":"1"}': ['5'],
        },
        Todo: {
          '5': {
            id: '5',
            title: 'finish collections',
          },
        },
        User: {
          '1': {
            id: '1',
            todos: '{"userId":"1"}',
            username: 'bob',
          },
        },
      },
      entityMeta: {
        'COLLECT:ArraySchema(Todo)': {
          '{"userId":"1"}': {
            date: 1557831718135,
            expiresAt: Infinity,
            fetchedAt: 0,
          },
        },
        Todo: {
          '5': {
            date: 1557831718135,
            expiresAt: Infinity,
            fetchedAt: 0,
          },
        },
        User: {
          '1': {
            date: 1557831718135,
            expiresAt: Infinity,
            fetchedAt: 0,
          },
        },
      },
      indexes: {},
      result: '1',
    };
    const state = normalize(
      [{ id: '10', title: 'create new items' }],
      User.schema.todos.push,
      [{ userId: '1' }],
      init.entities,
      init.indexes,
      init.entityMeta,
    );
    expect(state).toMatchSnapshot();
  });
});

describe(`${schema.Collection.name} denormalization`, () => {
  const normalizeNested = {
    entities: {
      'COLLECT:ArraySchema(Todo)': {
        '{"userId":"1"}': ['5'],
      },
      Todo: {
        '5': {
          id: '5',
          title: 'finish collections',
        },
      },
      User: {
        '1': {
          id: '1',
          todos: '{"userId":"1"}',
          username: 'bob',
        },
      },
    },
    entityMeta: {
      'COLLECT:ArraySchema(Todo)': {
        '{"userId":"1"}': {
          date: 1557831718135,
          expiresAt: Infinity,
          fetchedAt: 0,
        },
      },
      Todo: {
        '5': {
          date: 1557831718135,
          expiresAt: Infinity,
          fetchedAt: 0,
        },
      },
      User: {
        '1': {
          date: 1557831718135,
          expiresAt: Infinity,
          fetchedAt: 0,
        },
      },
    },
    indexes: {},
    result: '1',
  };

  test('denormalizes nested collections', () => {
    expect(
      denormalize(normalizeNested.result, User, normalizeNested.entities),
    ).toMatchSnapshot();
  });

  test('denormalizes top level collections', () => {
    expect(
      denormalize('{"userId":"1"}', userTodos, normalizeNested.entities),
    ).toMatchSnapshot();
  });

  describe('caching', () => {
    const entityCache = {};
    const resultCache = new WeakEntityMap();
    test('denormalizes nested and top level share referential equality', () => {
      const todos = denormalize(
        '{"userId":"1"}',
        userTodos,
        normalizeNested.entities,
        entityCache,
        resultCache,
        [{ userId: '1' }],
      );
      const user = denormalize(
        normalizeNested.result,
        User,
        normalizeNested.entities,
        entityCache,
        resultCache,
      );
      expect(user).toBeDefined();
      expect(user).not.toEqual(expect.any(Symbol));
      if (typeof user === 'symbol' || !user) return;
      expect(todos).toBe(user.todos);
    });

    test('push updates cache', () => {
      const pushedState = normalize(
        [{ id: '10', title: 'create new items' }],
        User.schema.todos.push,
        [{ userId: '1' }],
        normalizeNested.entities,
        normalizeNested.indexes,
        normalizeNested.entityMeta,
      );
      const todos = denormalize(
        '{"userId":"1"}',
        userTodos,
        pushedState.entities,
        entityCache,
        resultCache,
        [{ userId: '1' }],
      );
      const user = denormalize(
        normalizeNested.result,
        User,
        pushedState.entities,
        entityCache,
        resultCache,
        [{ id: '1' }],
      );
      expect(user).toBeDefined();
      expect(user).not.toEqual(expect.any(Symbol));
      if (typeof user === 'symbol' || !user) return;
      expect(user.todos.length).toBe(2);
      expect(todos).toBe(user.todos);
    });

    test('unshift places at start', () => {
      const unshiftState = normalize(
        [{ id: '2', title: 'from the start' }],
        User.schema.todos.unshift,
        [{ userId: '1' }],
        normalizeNested.entities,
        normalizeNested.indexes,
        normalizeNested.entityMeta,
      );
      const todos = denormalize(
        '{"userId":"1"}',
        userTodos,
        unshiftState.entities,
        entityCache,
        resultCache,
        [{ userId: '1' }],
      );
      const user = denormalize(
        normalizeNested.result,
        User,
        unshiftState.entities,
        entityCache,
        resultCache,
        [{ id: '1' }],
      );
      expect(user).toBeDefined();
      expect(user).not.toEqual(expect.any(Symbol));
      if (typeof user === 'symbol' || !user) return;
      expect(user.todos.length).toBe(2);
      expect(todos).toBe(user.todos);
      expect(user.todos[0].title).toBe('from the start');
    });

    test('denormalizes unshift', () => {
      const todos = denormalize(
        [{ id: '2', title: 'from the start' }],
        userTodos.unshift,
        {},
        entityCache,
        resultCache,
        [{ id: '1' }],
      );
      expect(todos).toBeDefined();
      expect(todos).not.toEqual(expect.any(Symbol));
      if (typeof todos === 'symbol' || !todos) return;
      expect(todos[0].title).toBe('from the start');
      expect(todos).toMatchInlineSnapshot(`
        [
          Todo {
            "completed": false,
            "id": "2",
            "title": "from the start",
            "userId": 0,
          },
        ]
      `);
    });

    test('denormalizes unshift (single item)', () => {
      const todos = denormalize(
        { id: '2', title: 'from the start' },
        userTodos.unshift,
        {},
        entityCache,
        resultCache,
        [{ id: '1' }],
      );
      expect(todos).toBeDefined();
      expect(todos).not.toEqual(expect.any(Symbol));
      if (typeof todos === 'symbol' || !todos) return;
      //expect(todos.title).toBe('from the start');
      expect(todos).toMatchInlineSnapshot(`
        {
          "id": "2",
          "title": "from the start",
        }
      `);
    });
  });
});
