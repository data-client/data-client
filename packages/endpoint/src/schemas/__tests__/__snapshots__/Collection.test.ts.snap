// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionSchema denormalization denormalizes nested collections 1`] = `
User {
  "email": "",
  "id": "1",
  "name": "",
  "todos": [
    Todo {
      "completed": false,
      "id": "5",
      "title": "finish collections",
      "userId": 0,
    },
  ],
  "username": "bob",
}
`;

exports[`CollectionSchema denormalization denormalizes top level collections 1`] = `
[
  Todo {
    "completed": false,
    "id": "5",
    "title": "finish collections",
    "userId": 0,
  },
]
`;

exports[`CollectionSchema normalization normalizes already processed entities 1`] = `
{
  "entities": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": [
        "5",
        "6",
      ],
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
  },
  "entityMeta": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "1",
}
`;

exports[`CollectionSchema normalization normalizes nested collections 1`] = `
{
  "entities": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": [
        "5",
      ],
    },
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
  },
  "entityMeta": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "1",
}
`;

exports[`CollectionSchema normalization normalizes push onto the end 1`] = `
{
  "entities": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": [
        "5",
        "10",
      ],
    },
    "Todo": {
      "10": {
        "id": "10",
        "title": "create new items",
      },
      "5": {
        "id": "5",
        "title": "finish collections",
      },
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
  },
  "entityMeta": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "Todo": {
      "10": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": [
    "10",
  ],
}
`;

exports[`CollectionSchema normalization normalizes top level collections 1`] = `
{
  "entities": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": [
        "5",
      ],
    },
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
    },
  },
  "entityMeta": {
    "COLLECT:ArraySchema(Todo)": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "{"userId":"1"}",
}
`;

exports[`CollectionSchema normalization should throw a custom error if data loads with string unexpected value 1`] = `
"Unexpected input given to normalize. Expected type to be "object", found "string".

          Schema: {
  "name": "Collection(Todo)",
  "schema": {
    "name": "Todo",
    "schema": {},
    "key": "Todo"
  },
  "key": "COLLECT:ArraySchema(Todo)"
}
          Input: "abc""
`;

exports[`CollectionSchema normalization should throw a custom error if data loads with string unexpected value 2`] = `
"Unexpected input given to normalize. Expected type to be "object", found "null".

          Schema: {
  "name": "Collection(Todo)",
  "schema": {
    "name": "Todo",
    "schema": {},
    "key": "Todo"
  },
  "key": "COLLECT:ArraySchema(Todo)"
}
          Input: "null""
`;

exports[`CollectionSchema normalization should throw a custom error if data loads with string unexpected value 3`] = `"Collections only work with @rest-hooks/react>=7.4"`;
